vortex-duckdb/src/copy.rs

```diff
@@ -0,0 +1,134 @@
+use std::fmt::Debug;
+use std::iter;
+
+use tokio::fs::File;
+use tokio::sync::mpsc;
+use tokio::sync::mpsc::Sender;
+use tokio::task::JoinHandle;
+use tokio_stream::wrappers::ReceiverStream;
+use vortex::ArrayRef;
+use vortex::dtype::Nullability::{NonNullable, Nullable};
+use vortex::dtype::{DType, StructFields};
+use vortex::error::{VortexExpect, VortexResult, vortex_err};
+use vortex::stream::ArrayStreamAdapter;
+use vortex_file::VortexWriteOptions;
+
+use crate::RUNTIME;
+use crate::convert::{data_chunk_to_arrow, from_duckdb_table};
+use crate::duckdb::{CopyFunction, DataChunk, LogicalType};
+
+#[derive(Debug)]
+pub struct VortexCopyFunction;
+
+pub struct BindData {
+    dtype: DType,
+    fields: StructFields,
+}
+
+pub struct GlobalState {
+    write_task: Option<JoinHandle<VortexResult<File>>>,
+    sink: Option<Sender<VortexResult<ArrayRef>>>,
+}
+
+impl CopyFunction for VortexCopyFunction {
+    type BindData = BindData;
+    type GlobalState = GlobalState;
+    type LocalState = ();
+
+    fn bind(
+        column_names: Vec<String>,
+        column_types: Vec<LogicalType>,
+    ) -> VortexResult<Self::BindData> {
+        let fields = from_duckdb_table(
+            column_names
+                .iter()
+                .zip(column_types)
+                .zip(iter::repeat(Nullable))
+                .map(|((name, type_), null)| (name, type_, null)),
+        )?;
+
+        Ok(BindData {
+            dtype: DType::Struct(fields.clone(), NonNullable),
+            fields,
+        })
+    }
+
+    fn copy_to_sink(
+        bind_data: &Self::BindData,
+        init_global: &mut Self::GlobalState,
+        _init_local: &mut Self::LocalState,
+        chunk: &mut DataChunk,
+    ) -> VortexResult<()> {
+        init_global
+            .sink
+            .as_ref()
+            .vortex_expect("sink closed early")
+            .blocking_send(data_chunk_to_arrow(bind_data.fields.names(), chunk))
+            .map_err(|e| vortex_err!("send error {}", e.to_string()))?;
+
+        Ok(())
+    }
+
+    fn copy_to_finalize(
+        _bind_data: &Self::BindData,
+        init_global: &mut Self::GlobalState,
+    ) -> VortexResult<()> {
+        RUNTIME.block_on(async {
+            if let Some(sink) = init_global.sink.take() {
+                drop(sink)
+            }
```

Comment:
Can we add some docs for the future reader here how the streaming copy roughly works. Without context, one wonders why tokio is looped in here.